plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -1174,10 +1178,31 @@ private void executeExpireSnapshots(ConnectorSession session, IcebergTableExecut
                 IcebergSessionProperties.EXPIRE_SNAPSHOTS_MIN_RETENTION);
 
         long expireTimestampMillis = session.getStart().toEpochMilli() - retention.toMillis();
+        TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+        List<String> pathsToDelete = new ArrayList<>();
+        Consumer<String> deleteFunction = path -> {
+            pathsToDelete.add(path);
+            if (pathsToDelete.size() == DELETE_BATCH_SIZE) {
```

Comment:
Is this check (and the deletion code associated to it) still needed?

Reply:
```
Tasks.foreach(manifestsToDelete)
        .executeWith(deleteExecutorService)
        .retry(3).stopRetryOn(NotFoundException.class).suppressFailureWhenFinished()
        .onFailure((manifest, exc) -> LOG.warn("Delete failed for manifest: {}", manifest, exc))
        .run(deleteFunc::accept)
```

From `org.apache.iceberg.RemoveSnapshots`.

You may end up in a situation where you add stuff with `Tasks.foreach`, but in another thread you reach `DELETE_BATCH_SIZE` and clear also paths that were not yet deleted.

Reply:
not true, current implementation in the worst case deletes file at the same time as the old one. The old one deletes files when `deleteFunc::accept` is called the new one either adds file to the list to be deleted later or deletes it at the time of he invocation.

Reply:
in a threaded context many things go wrong

- ArrayList can get corrupted, we may loose files
- `pathsToDelete.clear();` can be invoked with some paths never sent to deletion
- `deleteFiles` can be invoked twice on same/similar sets of paths and may result in "file missing so cannot delete" exception